File: sdk/typescript/test/test_client.mjs
import http from "node:http";
import { HyperionClient } from "../dist/index.js";

function createServer() {
  return http.createServer((req, res) => {
    const { method, url } = req;
    res.setHeader("Content-Type", "application/json");

    if (method === "GET" && url === "/api/monitoring") {
      res.end(JSON.stringify({ metrics: [] }));
    } else if (method === "GET" && url === "/api/health") {
      res.end(JSON.stringify({ active_replicas: 4 }));
    } else if (method === "GET" && url === "/api/deploy/status") {
      res.end(JSON.stringify({ last_state: "SUCCEEDED" }));
    } else if (method === "POST" && url === "/api/autoscale/plan") {
      let body = "";
      req.on("data", chunk => (body += chunk));
      req.on("end", () => {
        const payload = body ? JSON.parse(body) : {};
        const replicas = payload.replicas ?? 4;
        res.end(JSON.stringify({ decision: replicas < 6 ? "scale-up" : "no-op" }));
      });
      return;
    } else if (method === "POST" && url === "/api/deploy/plan") {
      let body = "";
      req.on("data", chunk => (body += chunk));
      req.on("end", () => {
        const payload = body ? JSON.parse(body) : {};
        res.end(JSON.stringify({ plan: ["scale", payload.desired_replicas ?? 4] }));
      });
      return;
    } else if (method === "POST" && url === "/api/deploy/apply") {
      res.end(JSON.stringify({ status: "applied" }));
    } else {
      res.statusCode = 404;
      res.end(JSON.stringify({ error: "not found" }));
    }
  });
}

async function run() {
  const server = createServer();
  await new Promise(resolve => server.listen(0, resolve));
  const { port } = server.address();

  const client = new HyperionClient({ baseUrl: `http://127.0.0.1:${port}`, apiKey: "test" });

  const monitoring = await client.getMonitoringSnapshot();
  if (!("metrics" in monitoring)) {
    throw new Error("Monitoring snapshot missing metrics field");
  }

  const autoscale = await client.autoscalePlan(3);
  if (autoscale.decision !== "scale-up") {
    throw new Error(`Unexpected autoscale decision: ${autoscale.decision}`);
  }

  const plan = await client.deploymentPlan({ desired_replicas: 8 });
  if (plan.plan[1] !== 8) {
    throw new Error("Deployment plan returned unexpected value");
  }

  await client.deploymentApply({ desired_replicas: 8 });
  await client.deploymentStatus();

  server.close();
}

run().catch(error => {
  console.error(error);
  process.exit(1);
});

File: sdk/python/tests/test_client.py
import json
import threading
import time
import unittest
from http.server import BaseHTTPRequestHandler, HTTPServer

from hyperion_sdk import HyperionClient


class _MockHandler(BaseHTTPRequestHandler):
    RESPONSES = {
        ("GET", "/api/monitoring"): ({"metrics": []}, 200),
        ("GET", "/api/health"): ({"active_replicas": 4}, 200),
        ("GET", "/api/deploy/status"): ({"last_state": "SUCCEEDED"}, 200),
    }

    def _write_json(self, payload, status=200):
        body = json.dumps(payload).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_GET(self):  # noqa: N802 (BaseHTTPRequestHandler naming)
        payload, status = self.RESPONSES.get(("GET", self.path), ({"error": "not found"}, 404))
        self._write_json(payload, status)

    def do_POST(self):  # noqa: N802
        length = int(self.headers.get("Content-Length", "0"))
        data = self.rfile.read(length).decode("utf-8") if length else "{}"
        payload = json.loads(data)
        if self.path == "/api/autoscale/plan":
            replicas = payload.get("replicas", 4)
            self._write_json({"decision": "scale-up" if replicas < 6 else "no-op"})
        elif self.path == "/api/deploy/plan":
            self._write_json({"plan": ["scale", payload.get("desired_replicas", 4)]})
        elif self.path == "/api/deploy/apply":
            self._write_json({"status": "applied"})
        else:
            self._write_json({"error": "unsupported"}, 404)

    def log_message(self, format, *args):  # noqa: A003 - silence test server logs
        return


class HyperionClientTests(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.server = HTTPServer(("127.0.0.1", 0), _MockHandler)
        cls.port = cls.server.server_port
        cls.thread = threading.Thread(target=cls.server.serve_forever, daemon=True)
        cls.thread.start()
        time.sleep(0.1)

    @classmethod
    def tearDownClass(cls):
        cls.server.shutdown()
        cls.thread.join(timeout=2)

    def setUp(self):
        self.client = HyperionClient(base_url=f"http://127.0.0.1:{self.port}", api_key="test")

    def test_monitoring_snapshot(self):
        snapshot = self.client.get_monitoring_snapshot()
        self.assertIn("metrics", snapshot)

    def test_autoscale_plan(self):
        decision = self.client.autoscale_plan(replicas=3)
        self.assertEqual(decision["decision"], "scale-up")

    def test_deployment_plan_and_apply(self):
        plan = self.client.deployment_plan({"desired_replicas": 6})
        self.assertEqual(plan["plan"], ["scale", 6])
        result = self.client.deployment_apply({"desired_replicas": 6})
        self.assertEqual(result["status"], "applied")


if __name__ == "__main__":  # pragma: no cover - direct execution support
    unittest.main()
